### PR TITLE
[SSDP] add `schema(Print &) const`

### PR DESCRIPTION
**Pull request by earlephilhower**
Supercedes #2806

Make SSDP::schema(WiFiClient&) use a by-ref (reduce stack use)

Add a SSDP::schema(Print&)

From @Palatis' original PR:
useful when using AsyncWebServer.

### CODE REVIEW

File: libraries/ESP8266SSDP/ESP8266SSDP.cpp [hunk 1]
```diff
@@ -295,6 +295,22 @@ void SSDPClass::schema(WiFiClient client) {
                );
 }
 
+void SSDPClass::schema(Print &print) const {
```
Comment (devyte) [inline libraries/ESP8266SSDP/ESP8266SSDP.cpp]: This looks very similar to the WiFiClient overload. Why have both? Can they be unified?

**Comment by earlephilhower** [comment]
@d-a-v, for the SSDP patch, I'm just the messenger.
I can't think of a good use case for a Printable vs. sending it over the wire, honestly.

However, maybe it makes sense to use just export `Print&` interface and drop the WiFiClient& one?  WiFiClient inherits a Print& interface from Stream...

**Comment by earlephilhower** [comment]
Err, @devyte  I meant to ping!

**Comment by devyte** [comment]
> maybe it makes sense to use just export Print& interface and drop the WiFiClient& one

Exactly why I asked. I suggest picking which of the two bodies looks best and keeping that one with Print arg.
Something that bothers me: the IP printout in both cases, one with the octets the other with IP2STR. I think the IPAddress::toString() method should be used instead, but that means updating the schema string, I think.

File: libraries/ESP8266SSDP/ESP8266SSDP.cpp (hunk 1 above)
Comment (earlephilhower) [reply libraries/ESP8266SSDP/ESP8266SSDP.cpp]: Done, only ::schema(Print&) survives.

**Comment by earlephilhower** [comment]
Both comments addressed.  SSDP will now use ip.toString() instead of octet extraction, and will export only the schema(Print&) which should work for any existing code as WiFiClient has a Print interface..

**Comment by earlephilhower** [comment]
Well, I guess C++ can't convert silently from a WiFiClient to Print for us.  Multiple inheritance C++ classes might have such limitations (not like Java where there's a first-class interface type).

Any ideas?

````
/home/travis/build/esp8266/Arduino/libraries/ESP8266SSDP/examples/SSDP/SSDP.ino:29:32: error: no matching function for call to 'SSDPClass::schema(esp8266webserver::ESP8266WebServerTemplate<WiFiServer>::ClientType)'

       SSDP.schema(HTTP.client());

                                ^

/home/travis/build/esp8266/Arduino/libraries/ESP8266SSDP/examples/SSDP/SSDP.ino:29:32: note: candidate is:

In file included from /home/travis/build/esp8266/Arduino/libraries/ESP8266SSDP/examples/SSDP/SSDP.ino:3:0:

/home/travis/arduino_ide/hardware/esp8266com/esp8266/libraries/ESP8266SSDP/ESP8266SSDP.h:65:10: note: void SSDPClass::schema(Print&) const

     void schema(Print &print) const;

          ^

/home/travis/arduino_ide/hardware/esp8266com/esp8266/libraries/ESP8266SSDP/ESP8266SSDP.h:65:10: note:   no known conversion for argument 1 from 'esp8266webserver::ESP8266WebServerTemplate<WiFiServer>::ClientType {aka WiFiClient}' to 'Print&'
````

**Comment by devyte** [comment]
Might be running into the 1-step implicit casting rule. I'll take a look at some point.

**Pull request merged**